text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,628 @@
+
+### Span names, kinds, and attributes
+
+#### Span name
+
+The span name should be descriptive and make it clear, what operation a span
+describes. In the context of messaging systems, this means that a span should
+at the very least make clear that it refers to a messaging system, in addition
+it needs to make clear what particular [messaging operation](#operation-name)
+it refers to.
+
+Ideally, the span name also contains the destination name of the messages it
+refers to. However, a destination name should only be added to the span name
+when it is of low cardinality. This is usually the case when the destination
+name is a meaningful and manually configured name (like a manually configure
+queue or topic name), it is usually not the case if the destination name is an
+auto-generated identifier (like a conversation id or an auto-generated name for
+an anonymous destination).
+
+> The span name SHOULD consist of the name of the messaging system followed by
+> an [operation name](#operation-name). The destination name MAY be appended if
+> it is of low cardinality.
+
+##### Examples
+
+* `kafka publish shop.orders`
+* `rabbitmq receive print_jobs`
+* `AmazonSQS deliver`
+* `activemq settle`
+
+#### Operation name
+
+The following operations related to messages are covered by these semantic
+conventions:
+
+| Operation name | Description |
+|----------------|-------------|
+| `publish`      | One ore more messages are provided for publishing to an intermediary. |
+| `create`       | A message is created. |
+| `receive`      | One or more messages are requested by a consumer. |
+| `deliver`      | One or more message are passed to a consumer. |
+| `settle`       | One or more message are settled. |
+
+For further details about each of those operations refer, to the [section about trace structure](#trace-structure).
+
+#### Span kind
+
+[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
+SHOULD be set according to the following table, based on the operation a span describes.
+
+| Operation name | Span kind|
+|----------------|-------------|
+| `publish`      | `PRODUCER`, if no `create` spans are present. `INTERNAL` otherwise. |
+| `create`       | `PRODUCER` |
+| `receive`      | `CONSUMER` |
+| `deliver`      | `CONSUMER` |
+| `settle`       | `INTERNAL` |
+
+Setting span kinds according to this table ensures, that span links between
+consumer and producers always go from a `PRODUCER` span on the producer side to
+a `CONSUMER` span on the consumer side. This allows analysis tools to interpret
+linked traces without the need of additional semantic hints.
+
+#### Attributes
+
+Attribute             | Type   | Required
+----------------------|--------|---------
+[`messaging.broker`](#messagingbroker) | string | Yes
+[`messaging.operation`](#messagingoperation) | string | Yes
+[`messaging.destination.name`](#messagingdestinationname) | string | For producer spans
```

Comment:
Based on a discussion in the workgroup, this attribute should fully and uniquely specify the tenant and queue/topic.

Reply:
We will come with a list of how `destination.name` might look for different messaging systems and brokers.

Reply:
To me it feels a bit arbitrary to include tenant but not broker in `destination.name`. The (broker, tenant, queue/topic) tuple uniquely defines a destination. If the goal is for `destination.name` specify a unique destination, then it should include all 3 components, not just tenant and queue/topic.

However, IMO it is actually better to split up individual pieces of information into different attributes to allow for more flexible searching in back ends. So, I think I'd prefer tenant to be split into its own attribute. I'm not sure if it's better off as an attribute of the destination or as a `messaging.tenant`. Perhaps we can discuss further next week?

Reply:
For moving forward with this discussion, I looked at different messaging systems and what is needed to identify a destination for those. I then tried to map this information to the _messaging.destination_ and _messaging.url_ attributes in the current specification.

### Kafka
* Servers, e. g. `host1:port1,host2:port2` (_messaging.url_)
* Topic name (_messaging.destination_)

### JMS
* ConnectionFactory
* Destination, either aqueue or a topic name (_messaging.destination_)

### RabbitMQ
* Server (_messaging.url_)
* Exchange (_messaging.destination_)
* Routing key (_messaging.destination_)

### NATS
* Servers, e. g. `nats://server1:port1,nats://server2:port2` (_messaging.url_)
* Subject (_messaging.destination_)

### Pulsar
* Server, e. g. `pulsar://localhost:6650` (_messaging.url_)
* Topic (_messaging.destination_)

### RocketMQ
* Server, e. g. `localhost:9876` (_messaging.url_)
* Topic (_messaging.destination_)

### Amazon SQS
* Region
* Account number
* Queue name (_messaging.destination_)

All is embedded in a Queue URL (_messaging.url_): `https://sqs.us-east-2.amazonaws.com/123456789012/MyQueue`

### Amazon SNS
* Region
* Account number
* Topic name (_messaging.destination_)

All is embedded in a topic ARN (_messaging.url_?): `arn:aws:sns:us-east-2:123456789012:MyTopic`

### Amazon Kinesis
* Region
* Account number
* Stream name  (_messaging.destination_)

All is embedded in a stream ARN (_messaging.url_?): `arn:aws:kinesis:us-west-2:123456789012:stream/Foo`

### Azure Event Hubs
* Event Hub namespace
* Event Hub name (_messaging.destination_)

The namespace is embedded in a connection string (_messaging.url_): `Endpoint=sb://<NamespaceName>.servicebus.windows.net/;SharedAccessKeyName=<KeyName>;SharedAccessKey=<KeyValue>;EntityPath=<EventHubName>`

### Azure Service Bus
* Service bus namespace
* Service bus queue  (_messaging.destination_)

The namespace is embedded in a connection string (_messaging.url_): `Endpoint=sb://<NamespaceName>.servicebus.windows.net/;SharedAccessKeyName=<KeyName>;SharedAccessKey=<KeyValue>;EntityPath=<EventHubName>`

### Google Pub/Sub
* Project id
* Topic id (_messaging.destination_)

All is embedded in a resource name (_messaging.url_?): `projects/sampleproject/topics/sampletopic`

### AMQP
* Server address, e. g. `amqp://guest:guest@localhost:5672`  (_messaging.url_)
* Node name (_messaging.destination_)

### MQTT
* Server, e. g. `localhost:5555`  (_messaging.url_)
* Topic name (_messaging.destination_)

Reply:
One proposed solution from the last workgroup meeting:

* `messaging.destination.name` as required attribute, capturing the topic, queue, subject, or node name, as previously.
* `messaging.destination.url` as an optional attribute, capturing the URL, or server name.

In addition to that, messaging systems can extend this with their own (possibly required) attributes. E. g., Azure Event Hubs could have `messaging.destination.name` and `messaging.destination.namespace` both required, Amazon solutions could require a combination of `messaging.destination.name`, `messaging.destination.account`, and `messaging.destination.region`.

Reply:
I agree with the general approach, but I have a couple of comments.

> In addition to that, messaging systems can extend this with their own (possibly required) attributes.
Do we want messaging systems to extend this to their liking, directly inside of `messaging.destination`? Or would it be better to suggest that these be scoped to `messaging.destination.<system>.<system-specific-attribute>`. This makes things a bit less confusing, I think, if different systems end up with the same attribute but with different format and/or semantics in the different systems.

> `messaging.destination.url` as an optional attribute, capturing the URL, or server name.
Since the `url` is optional, I don't see a problem per se. But one thing I'd like us to be aware of is that for consumers, the required grouping of attributes would be `messaging.source` and `messaging.destination` would be optional, as some messaging systems may provide some information related to the message's original destination and others may not. For systems that do have this information available, in many cases there would be much less information available for the original destination compared to what the producer of the message had. In some cases, only the destination name would be known, but not necessarily the URL and/or other information to uniquely identify the message's original destination. In the case of `url`, it might be different if the receiving broker was part of a network and forwarded the message to another broker before being sent to a consumer. These two brokers would have different URLs.

In cases where the consumer can provide partial information of the original destination, but not enough to uniquely identify the destination, is it better to provide the partial information as a clue, even if the destination can't be uniquely identified?

Reply:
> In cases where the consumer can provide partial information of the original destination, but not enough to uniquely identify the destination, is it better to provide the partial information as a clue, even if the destination can't be uniquely identified?

I would say so. If we stick with having both `messaging.source` and `messaging.destination` (which we didn't finish discussing yet), for consumers all `messaging.destination.*` attributes would be optional, and I think we shouldn't require any combination to be present that lets one uniquely identify the source. This would complicate the specification without providing much benefit.

Reply:
@dpauls the formatting was a bit confusing, but I think this was your comment/question right?

>it be better to suggest that these be scoped to messaging.destination.<system>.<system-specific-attribute>. This makes things a bit less confusing, I think, if different systems end up with the same attribute but with different format and/or semantics in the different systems.

I think that's generally what @pyohannes meant when saying 

>In addition to that, messaging systems can extend this with their own (possibly required) attributes.

Just wanted to make sure we are all in the same page here :)

Reply:
We still didn't finish this topic, so just leaving this comment to make the status clear.